File: src/session-order.ts
import fs from "node:fs";
import path from "node:path";
import os from "node:os";

const ORDER_FILE = path.join(os.homedir(), ".muxtunnel", "session-order.json");

let cachedOrder: string[] = [];

export function loadSessionOrder(): string[] {
  try {
    const raw = fs.readFileSync(ORDER_FILE, "utf-8");
    const parsed = JSON.parse(raw);
    if (Array.isArray(parsed) && parsed.every((s) => typeof s === "string")) {
      cachedOrder = parsed;
    } else {
      cachedOrder = [];
    }
  } catch {
    cachedOrder = [];
  }
  return cachedOrder;
}

export function getSessionOrder(): string[] {
  return cachedOrder;
}

export function saveSessionOrder(order: string[]): void {
  cachedOrder = order;
  try {
    fs.mkdirSync(path.dirname(ORDER_FILE), { recursive: true });
    fs.writeFileSync(ORDER_FILE, JSON.stringify(order, null, 2) + "\n");
  } catch (err) {
    console.error("[session-order] Failed to save:", err);
  }
}

File: src-tauri/src/session_order.rs
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;

static ORDER: once_cell::sync::Lazy<Mutex<Vec<String>>> =
    once_cell::sync::Lazy::new(|| Mutex::new(vec![]));

fn order_file() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_default()
        .join(".muxtunnel")
        .join("session-order.json")
}

pub fn load() {
    let path = order_file();
    let order = match fs::read_to_string(&path) {
        Ok(raw) => match serde_json::from_str::<Vec<String>>(&raw) {
            Ok(v) => v,
            Err(_) => vec![],
        },
        Err(_) => vec![],
    };
    *ORDER.lock().unwrap() = order;
}

pub fn get() -> Vec<String> {
    ORDER.lock().unwrap().clone()
}

pub fn save(order: Vec<String>) {
    *ORDER.lock().unwrap() = order.clone();
    let path = order_file();
    if let Some(parent) = path.parent() {
        let _ = fs::create_dir_all(parent);
    }
    if let Err(e) = fs::write(&path, serde_json::to_string_pretty(&order).unwrap_or_default()) {
        log::error!("[session-order] Failed to save: {}", e);
    }
}
